chain/emitter/app.go

```diff
@@ -200,32 +211,39 @@ func (app *App) DeliverTx(req abci.RequestDeliverTx) abci.ResponseDeliverTx {
 			"extra": extra,
 		})
 	}
+	app.AddAccountsInTx(stdTx.GetSigners()[0])
+	accsInTx := []sdk.AccAddress{}
+	for accStr, _ := range app.accsInTx {
+		acc, _ := sdk.AccAddressFromBech32(accStr)
+		accsInTx = append(accsInTx, acc)
+	}
+
+	txDict["related_accounts"] = accsInTx
+	app.AddAccountsInBlock(accsInTx...)
+	app.accsInTx = make(map[string]bool)
 	txDict["messages"] = messages
-	app.AddAccounts(stdTx.GetSigners()[0])
 	return res
 }
 
 // EndBlock calls into the underlying EndBlock and emits relevant events to Kafka.
 func (app *App) EndBlock(req abci.RequestEndBlock) abci.ResponseEndBlock {
 	res := app.BandApp.EndBlock(req)
-	// Update balances of all affected accounts on this block.
-	accMap := make(map[string]bool)
-	for _, acc := range app.accs {
-		accStr := string(acc)
-		if accMap[accStr] {
-			continue
-		}
-		accMap[accStr] = true
-		app.Write("SET_ACCOUNT", JsDict{
-			"address": acc,
-			"balance": app.BankKeeper.GetCoins(app.DeliverContext, acc).String(),
-		})
-	}
-
 	for _, event := range res.Events {
 		app.handleBeginBlockEndBlockEvent(event)
 	}
-
+	// Update balances of all affected accounts on this block.
+	msgs := []Message{}
+	for accStr, _ := range app.accsInBlock {
+		acc, _ := sdk.AccAddressFromBech32(accStr)
+		msgs = append(msgs, Message{
+			Key: "SET_ACCOUNT",
+			Value: JsDict{
+				"address": acc,
+				"balance": app.BankKeeper.GetCoins(app.DeliverContext, acc).String(),
+			}})
+	}
+	// Index 0 is message new block, We insert set account message after new block message
```

Comment:
// Index 0 is message **NEW_BLOCK**, **w**e insert **SET_ACCOUNT** message**s** right after it.